### PR TITLE
Missing memcache should not cause occ hard-fail

### LINKED ISSUE #17329: Missing memcache class on CLI upgrade
### Steps

1. Setup OC 8.1RC1
2. Enable APCu memcache (enable the php module and set `'memcache.local' => '\OC\Memcache\APCu'` in config.php
3. Go to admin page, see that there is no warning about memcache
4. Switch code to OC 8.1RC2
5. Run CLI upgrade with `sudo -uwwwrun ./occ upgrade`

### Expected result
Upgrade works

### Actual result
```
An unhandled exception has been thrown:
OC\HintException: [0]: Missing memcache class \OC\Memcache\APCu for local cache (Is the matching PHP module installed and enabled ?)
```
Please note that the web UI upgrade works properly.

### Versions
ownCloud 8.1 RC2

@icewind1991 @DeepDiver1975

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
Warning is now printed to logs, but occ will still work. The memcache factory is primed with `null` memcaches in this case, so no caching will be performed.

Fixes #17329 

cc @MorrisJobke @DeepDiver1975 @RealRancor @rullzer @MightyCreak

BTW, I noticed something a little strange... in case the ownCloud instance isn't installed yet, or debug is enabled, we fall back to using ArrayCache, but the Factory is constructed using ArrayCache instances, while the factory itself needs class names (it initialises the caches itself when required). I think this is a bug? cc @LukasReschke

### CODE REVIEW

**Comment by RealRancor** [comment]
Can't say something about the code but is this also catching the issue where cron.php is not running because of the missing APC/APCu?

File: lib/private/server.php [hunk 1]
```diff
@@ -234,11 +234,23 @@ public function __construct($webRoot) {
...
-					$config->getSystemValue('memcache.locking', null)
-				);
+				try {
+					return new \OC\Memcache\Factory($prefix,
+						$config->getSystemValue('memcache.local', null),
+						$config->getSystemValue('memcache.distributed', null),
+						$config->getSystemValue('memcache.locking', null)
+					);
+				} catch (HintException $e) {
+					if (defined('OC_CONSOLE')) {
```
Comment (MorrisJobke) [inline lib/private/server.php]: Use `\OC::$CLI` instead and it will fix the cron.php problem too :)

**Comment by MorrisJobke** [comment]
I tested this and with the proposed change it will also fix the cron stuff. Maybe adjust the log message then to CLI instead of occ

File: lib/private/server.php [hunk 2]
```diff
@@ -234,11 +234,23 @@ public function __construct($webRoot) {
...
+						// CLI should not hard-fail on broken memcache
+						$c->getLogger()->warning('Memcache disabled for occ: {msg}', [
```
Comment (MorrisJobke) [inline lib/private/server.php]: info is enough for this in my opinion

File: lib/private/server.php (hunk 1 above)
Comment (Xenopathic) [inline lib/private/server.php]: Ah, that's the variable I was looking for!

**Comment by Xenopathic** [comment]
Screw you too Jenkins

@owncloud-bot retest this please

**Comment by MightyCreak** [comment]
Since it's not recommended to enable memcache with APCu on CLI, why don't we simply test directly in the factory?

Something like that:
```php
if (\OC::$CLI && $memcache === "APCu") {
    // APCu memcache is not recommended by upstream for CLI.
    return null;
}
else {
    // all good.
    return new Memcache;
}

}
```

**Comment by Xenopathic** [comment]
@MightyCreak The problem here isn't specific to APCu, XCache will also have the same problem (it runs in-process too). What's more, the CLI should never *need* a memcache for operation. The only reason the exception throwing was introduced was due to the idiom that any unavailable service should prevent the OC instance from running, be that LDAP, DB or (now) memcache. For CLI this isn't a problem, so memcache failures shouldn't be a hard-fail.

**Comment by MightyCreak** [comment]
@Xenopathic Ok, but then why should we catch an exception from \OC\Memcache\Factory? The `if`in the catch should directly be in the Factory function, don't you think?

Catching an exception to handle a case that is not an error seems like a workaround to me. But if everyone is fine with that, then it's fine by me ;)

**Comment by Xenopathic** [comment]
@MightyCreak Hmm, that might be a better idea. Avoiding `catch` is always a good thing!

Now get back to Mesamatrix and write new features :laughing:

**Comment by MightyCreak** [comment]
Sir, yes sir! ^^

**Comment by Xenopathic** [comment]
So I moved the checking logic to the memcache factory, removing the `try catch` statements. Also modified the wording of the exception text/logging text:

> Memcache {class} not available for {use} cache

instead of 

> Missing memcache class {class} for {use} cache

Also pre-emptively fixed the issue I mentioned in the original post, with the class name vs object.

In case this is worse than what was here before, I'm leaving this commit ID to revert back to: a629f16cacc55227f769acde3c9c46a24a36c600

**Comment by Xenopathic** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MightyCreak** [comment]
:+1: ^^

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
Jenkins PR: #17681

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by MorrisJobke** [comment]
@karlitschek Should we backport this to fix the broken behaviour of APCu on CLI for the occ tool? The original ticket was against 8.1.1

**Comment by karlitschek** [comment]
I think a backport would be good :+1:

**Comment by MorrisJobke** [comment]
@Xenopathic Can you please backport? Keep in mind: stable8.1 freeze is planned for monday evening ... ;)

**Pull request merged**